### PR TITLE
Deposit form

### PR DESCRIPTION
**Pull request by anxolin**
image

**Includes**
- [X] Show/hide both Deposit form and Withdraw form
- [X] Close button functionality: Closes and clear the form
- [X] Markup and CSS for the form
- [X] Add a "Use max" link that will use all the available balance
- [X] Form validation: Validates that the user input is a number and in the right range

**Utils**
Aditionally improved and added some aditional utils:
- [X] Format amounts: Allow to remove thousands, so we can format amounts for inputs too
- [X] Format amounts: Handle edge cases, like the one where the coin has less precission than the decimals we want to show. Also add more tests
- [X] Parse amounts: Required to convert the decimal amount that the user inputs into a BN for the selected currency. Added unit tests for this function

**Bonus :)**
- [X] Applied a refactor between the Deposit and Withdraw form. This way, we can use the same component for both
- [X] Added already the two forms for deposit and withdraw with their own parametrization 

**Do not include**
It doesn't include the call to the service, update of the balances, or rollback in case of error.

**Test**:
* `yarn test`
* Check link: https://pr76--dexreact.review.gnosisdev.com

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr76--dexreact.review.gnosisdev.com

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr76--dexreact.review.gnosisdev.com

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr76--dexreact.review.gnosisdev.com

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr76--dexreact.review.gnosisdev.com

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr76--dexreact.review.gnosisdev.com

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr76--dexreact.review.gnosisdev.com

File: src/components/DepositWidget/Form.tsx [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    if (validatorActive) {
+      // Verify on every deposit change
+      const errorMsg = _validateDeposit(totalAmount, depositAmount, decimals)
+      const newErrors = {
+        ...errors,
+        depositAmount: errorMsg,
+      }
+      setErrors(newErrors)
+      // const hasErrors = Object.keys(newErrors).some(key => !!newErrors[key])
+      // return !hasErrors
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/components/DepositWidget/Form.tsx]: Yes, I'll remove them. Although, I was not very happy with how I solved that it was validating the form before sending it, and it was weird.

There's a TODO somewhere related to that, I might give it another thought when I address the call to the service...

File: src/components/DepositWidget/Form.tsx [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  submitBtnIcon: IconDefinition
+  onSubmit: () => Promise<void>
+  onClose: () => void
+}
+
+interface Errors {
+  depositAmount: string
+}
+
+function _validateDeposit(totalAmount: BN, depositAmount: string, decimals: number): string {
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/components/DepositWidget/Form.tsx]: Good catch! I did a refactor bu I overlook that. `totalAmount` was `walletAmount` before, and it was refactored. I forgot that one...

File: src/utils/format.ts [hunk 3]
```diff
@@ -41,31 +40,58 @@ export function formatAmount(
...
-  return formatAmount(amount, amountPrecision, amountPrecision)
+  return formatAmount(amount, amountPrecision, amountPrecision, thousandSeparator)
+}
+
+export function parseAmount(amountFmt: string, amountPrecision = DEFAULT_PRECISION): BN | null {
+  if (!amountFmt) {
+    return null
+  }
+
+  const groups = /^(\d+)(?:\.(\d+))?$/.exec(amountFmt)
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/utils/format.ts]: Cause BN is only for integers. It doesn't suport decimals. Users enter the amounts with decimals, because they don't work in weis. BN.js works in weis.

Maybe I'm overlooking something, because I also thought it was weird there was not more libs out there doing formatting/parsing of BN.js

Let me know if you can think of a better way

File: test/utils/formatAmount.spec.ts [hunk 4]
```diff
@@ -82,13 +86,52 @@ describe('Tokens with precision 6', () => {
...
+  })
+
+  test('12,345 units', async () => {
+    const amount = new BN('12345')
+    expect(formatAmount(amount, 0)).toEqual('12,345')
+  })
+
+  test('4,567,890 units, round down', async () => {
+    const amount = new BN('4567890')
+    expect(formatAmount(amount, 0)).toEqual('4,567,890')
```
[the remark this answers is not in the document]
Comment (anxolin) [reply test/utils/formatAmount.spec.ts]: U are right. C&P :) 
Removing...

File: src/utils/format.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/utils/format.ts]: Ahhh got it. Makes sense now :)

File: src/components/DepositWidget/Form.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/components/DepositWidget/Form.tsx]: Have you changed it locally and forgot to push? I see no new changes in this PR.

File: test/utils/formatAmount.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply test/utils/formatAmount.spec.ts]: Same here, it's still there

File: src/components/DepositWidget/Form.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (anxolin) [reply src/components/DepositWidget/Form.tsx]: image

Upsssss...

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr76--dexreact.review.gnosisdev.com
